File: full-data-structures/Data_structures/QuickSort/quicksort.js
function quicksort(arr) {
    if (arr.length <= 1) return arr;

    const pivot = arr[arr.length - 1];
    const left = [];
    const right = [];

    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] < pivot) {
            left.push(arr[i]);
        } else {
            right.push(arr[i]);
        }
    }

    return [...quicksort(left), pivot, ...quicksort(right)];
}

console.log(quicksort([8, 3, 1, 7, 0, 10, 2]))

//time complexity best case -> O(n log n);
//time compelxity worst case -> O(n^2);
//space complexity: O(log n)

File: quick_sort_py/quicksort.py
def quicksort(arr):
    if len(arr) < 2:
        return arr
    else:
        pivot = arr[0]
        less = [i for i in arr[1:] if i <= pivot]
        greater = [i for i in arr[1:] if i > pivot]

        return quicksort(less) + [pivot]  + quicksort(greater)
    

if __name__ == "__main__":
    sample_array = [10, 5, 2, 3]
    print("Unsorted array:", sample_array)
    sorted_array = quicksort(sample_array)
    print("Sorted array:", sorted_array)

# time complexity O(n log n) on average, O(n^2) in the worst case
# space complexity O(log n) due to recursive stack space

